Mathlib/Combinatorics/Flow/Basic.lean

```diff
@@ -0,0 +1,232 @@
+/-
+Copyright (c) 2024 Niklas Mohrin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Niklas Mohrin
+-/
+import Mathlib.Algebra.BigOperators.Group.Finset
+import Mathlib.Combinatorics.Digraph.Basic
+
+/-!
+# Network Flows
+
+In this file we introduce network flows. A flow network can be seen as a directed graph where each
+arc of the graph has a fixed capacity. A flow from a source vertex `s` to a sink vertex `t` is an
+assignment of "flow units" to each edge, such that for all vertices other than the source and the
+sink, the incoming flow must equal the outgoing flow (_flow conservation_). The value of such an
+assignment is the amount of flow units leaving the source `s` and arriving at the sink `t`. In
+practice, we are usually interested in finding a flow of maximal value for a given network.
+
+## Main definitions
+
+- `Network`: A flow network assigning capacities to all pairs of vertices. Notably, capacities and
+             non-negative and the capacity from a vertex to itself is zero.
+- `PseudoFlow`, `PreFlow`, and `Flow`: An assignment of flow values for a given `Network`, abiding
+                                       increasingly strict constraints.
+
+## Notation
+
+For flows `f f' : N.PseudoFlow`, we define `f ≤ f'` as the _subflow relation_, indicating that for
+all `u v : V`, `f u v ≤ f' u v`.
+
+## References
+
+See [goldberg_network_1989] for a thorough introduction to the topic.
+
+## Tags
+
+flow, network, max-flow, maxflow, cut, min-cut, mincut, graph, subflow
```

Comment:
Do we really need two different tags `maxflow` and `max-flow`?

Reply:
same for `mincut` and `min-cut`

Reply:
I don't know, @YaelDillies ?